doc/sphinx/language/cic.rst

```diff
@@ -111,51 +111,49 @@ Typing rules
 ----------------
 
 As objects of type theory, terms are subjected to *type discipline*.
-The well typing of a term depends on a global environment and a local
-context.
-
+The well typing of a term depends on a local context and a global environment.
 
 .. _Local-context:
 
 **Local context.**
-A *local context* is an ordered list of *local declarations* of names
-which we call *variables*. The declaration of some variable :math:`x` is
-either a *local assumption*, written :math:`x:T` (:math:`T` is a type) or a *local
-definition*, written :math:`x:=t:T`. We use brackets to write local contexts.
-A typical example is :math:`[x:T;~y:=u:U;~z:V]`. Notice that the variables
+A :term:`local context` is an ordered list of declarations of *variables*.
```

Comment:
I'm thinking it's a bit weird to have a full paragraph here but to link to a much shorter explanation where the `:gdef:` is in the proof mode chapter.
It would make much more sense to accept that these two paragraphs *are* the definitions of "local context" and "global environment".
In this case, the two terms could be introduced with `.. glossary::` since they are already separated in paragraphs:

```rst
.. glossary::

   local context

     A local context is an ordered list of declarations of local :gdef:`variables <variable>`...

   global environment

     A global environment is an ordered list of declarations of global :gdef:`constants <constant>`...
```

Ideally, we should even move these two glossary definitions to the "Essential vocabulary" subsection (of the "Basic notions and conventions" section of the "Core language" chapter) but this would require more work because the language would need to be adapted further.

Reply:
I think users will more frequently want to find the shorter, more readable definition in proof-mode that takes the perspective of someone who's trying to use a piece of software.  The Cic chapter is mathematical and introduces a lot of notation that's (mostly?) only used in that chapter.  That will less frequently be of interest to general users.  I would keep the glossary pointing to here.  I'll add links to the Cic chapter so that it is discoverable.

At some time I'd like to harmonize the Cic and proof-mode language related to local context/global environment, but not today.

Reply:
Note the CiC definition of local context is only one sentence, followed by a lot of detail on the mathematical notation.  The definition of global environment in CiC (only 3 sentences) doesn't explain what an assumption, definition or declaration is, either, though it give the notations for the first two in detail.  So could use some work.

Reply:
> The Cic chapter is mathematical and introduces a lot of notation that's (mostly?) only used in that chapter. That will less frequently be of interest to general users. I would keep the glossary pointing to here. I'll add links to the Cic chapter so that it is discoverable.

I agree with this. Note that this is why I said that ideally we should move these definitions to the "Essential vocabulary" but that would require rewording.

>So could use some work.

I'll see what I can do about this after this PR is merged.

Reply:
I think we need a several-page overall introduction before the Core Language chapter that introduces the concepts and all of Coq's capabilities in somewhat conversational prose (like what we're adding in the proof-mode and tactics chapters in the PR, though longer).  I could take a crack at writing something along those lines.  I think the key is explaining sufficiently to allow easy understanding of the relationships between concepts.  The structure should be pyramidal--the overall introduction gives you the big picture while omitting most of the details.  Then chapter/section introductions drill down some more, and then below that is full detail, commands, tactics, etc.

The first mention of a concept should be accurate but it doesn't have to be the most detailed explanation.  If you have hyperlinks, it's not necessary to group a lot of definitions in a single section.